src/charm.py

```diff
@@ -39,10 +41,124 @@ def __init__(self, *args):
 
         self.framework.observe(self.on.install, self._on_install)
         self.framework.observe(self.on.leader_elected, self._on_leader_elected)
+        self.framework.observe(self.on.get_primary_action, self._on_get_primary)
         self.framework.observe(self.on.start, self._on_start)
         self.framework.observe(self.on.get_initial_password_action, self._on_get_initial_password)
         self._cluster = Patroni(self._unit_ip)
 
+    def _on_get_primary(self, event: ActionEvent) -> None:
+        """Get primary instance."""
+        try:
+            primary = self._patroni.get_primary(unit_name_pattern=True)
+            event.set_results({"primary": primary})
+        except RetryError as e:
+            logger.error(f"failed to get primary with error {e}")
+
+    @retry(
+        stop=stop_after_delay(60),
+        wait=wait_fixed(5),
+        reraise=True,
+    )
+    def _change_primary(self) -> None:
+        """Change the primary member of the cluster."""
+        # Inform the first of the remaining available members to not incur the risk
+        # of triggering a switchover to a member that is also being removed.
+        if not self._patroni.are_all_members_ready():
+            logger.info("could not switchover because not all members are ready")
```

Comment:
maybe defer to wait until all members are ready?

Reply:
Or put this test on the calling handler, since there's no access to the event from the function.

Reply:
The event should not be deferred because it is the storage detaching event. I moved this check to the calling handler in the next PR and removed it from here in https://github.com/canonical/postgresql-operator/pull/10/commits/b92efb42f7a7fd1e79bfd418afae2e9a828fa46f.